esx_service/vmdk_ops.py

```diff
@@ -252,31 +253,62 @@ def getVMDKUuid(vmdk_path):
     except:
         return None
 
+# Create and return the devFS path for a VSAN object UUID.
+# Also returns true if clean up is needed.
+# cleanup_vsan_devfs should be called to clean up before
+# removing the VSAN object.
+def get_vsan_devfs_path(uuid):
+
+    logging.debug("Got volume UUID %s", uuid)
+
+    # Objtool creates a link thats usable to 
+    # read write to vsan object.
+    cmd = "{0} {1}".format(OBJ_TOOL_CMD, uuid)
+    rc, out = RunCommand(cmd)
+    if rc == 0 and os.path.isfile(fpath):
+        return "/vmfs/devices/vsan/{0}".format(uuid), True
+    return None, False
+
+# Clean up vsan devfs path
+def cleanup_vsan_devfs_path(devfs_path)
+    try:
+        os.remove(backing)
+        logging.debug("Unlinked %s for flag %s", backing, cleanup_device)
+	return True
+    except OSError as ex:
+        logging.error("Failed to remove backing device %s",
+                      backing, str(ex))
+    return False
+
+# Returns the flat file for a VMDK.
+def get_backing_flat_file(vmdk_path):
+    return vmdk_path.replace(".vmdk", "-flat.vmdk")
 
 # Return a backing file path for given vmdk path or none
-# if a backing can't be found.
-def getVMDKBacking(vmdk_path):
-    flatBacking = vmdk_path.replace(".vmdk", "-flat.vmdk")
+# if a backing can't be found. Returns True if clean up 
+# is needed. Do not forget to call cleanup_backing_device when done.
+def get_backing_device(vmdk_path):
+    flatBacking = get_backing_flat_file(vmdk_path)
     if os.path.isfile(flatBacking):
-        return flatBacking
+        return flatBacking, False
 
-    uuid = getVMDKUuid(vmdk_path)
+    uuid = get_vsan_uuid(vmdk_path)
 
     if uuid:
-        logging.debug("Got volume UUID %s", uuid)
-        # Objtool creates a link thats usable to format the
-        # vsan object.
-        cmd = "{0} {1}".format(OBJ_TOOL_CMD, uuid)
-        rc, out = RunCommand(cmd)
-        fpath = "/vmfs/devices/vsan/{0}".format(uuid)
-        if rc == 0 and os.path.isfile(fpath):
-            return fpath
-    return None
 
+        return get_vsan_devfs_path(uuid)
+
+    return None, False
+
+def cleanup_backing_device(backing, cleanup_device):
+    if cleanup_device:
+    	return cleanup_vsan_devfs_path(backing)
+    return True
```

Comment:
Cleanup is used as a symmetric function for get, so this would be a nice check but ideally done only for debug. Also, once this code moves into VM these functions will no longer be present.